File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter11/Petzold.ProgrammingWindows6.Chapter11/Petzold.ProgrammingWindows6.Chapter11/ColorToContrastColorConverter.cs
using System;
using Windows.UI;
using Windows.UI.Xaml.Data;

namespace Petzold.ProgrammingWindows6.Chapter11
{
    public class ColorToContrastColorConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, string language)
        {
            Color clr = (Color)value;
            double grayShade = 0.30 * clr.R + 0.59 * clr.G + 0.11 * clr.B;
            return grayShade > 128 ? Colors.Black : Colors.White;
        }
        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            return value;
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter11/PetzoldProgrammingWindows6Chapter11/PetzoldProgrammingWindows6Chapter11/ColorToContrastColorConverter.cpp
#include "pch.h"
#include "ColorToContrastColorConverter.h"

using namespace PetzoldProgrammingWindows6Chapter11;

using namespace Platform;
using namespace Windows::UI;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Interop;

Object^ ColorToContrastColorConverter::Convert(Object^ value, TypeName targetType, 
                                               Object^ parameter, String^ langauge)
{
    Color clr = (Color)value;
    double grayShade = 0.3 * clr.R + 0.59 * clr.G + 0.11 * clr.B;
    return grayShade > 128 ? Colors::Black : Colors::White;
}

Object^ ColorToContrastColorConverter::ConvertBack(Object^ value, TypeName targetType, 
                                                   Object^ parameter, String^ language)
{
    return value;
}